### PR TITLE
Fix: add lock to prevent repeating Name File dialogs

### PR DESCRIPTION
**Pull request by jess-x**
## References
Fix https://github.com/jupyterlab/jupyterlab/issues/10363 &  https://github.com/jupyterlab/jupyterlab/issues/10291

Edit: I am breaking https://github.com/jupyterlab/jupyterlab/pull/10294 up, hopefully that makes things easier to test 

## Code changes
Added `activatedDocs` used as a naive implemetation of lock for Name File

## User-facing changes
Repeating Name File dialogs should be removed

## Backwards-incompatible changes

None I am aware of

### CODE REVIEW

File: packages/docmanager-extension/src/index.ts [hunk 1]
```diff
@@ -804,25 +804,37 @@ function addCommands(
   });
 
   const newFileRegex = new RegExp('^untitled', 'i');
-
+  /**
+   * An array of activated file names used as a naive implemetation of lock for Name File
```
Comment (echarles) [inline packages/docmanager-extension/src/index.ts:808]:
```suggestion
   * An array of activated file names used as a naive implementation of lock for Name File
```

**Comment by echarles** [comment]
Thx @jess-x.

To better understand the fix, can you detail the failing use case that you are fixing with this?

Question: as the list of files is store in an array `activatedDocs`, what will happen if the user refreshes (reloads) his page?

**Comment by jess-x** [comment]
Thanks @echarles, 

the failing use case intended to fix here: where multiple `activateRequested` signals are emitted simultaneously, multiple `nameOnSaveDialog`s are returned simultaneously, hence the repeated popups. 

Adding a lock around the return statement should only allow only one `nameOnSaveDialog` to return each time the saveState turns `completed manually`, either by 'Ctrl-s' or clicking the save button. That said, I also looked at if there is a better option than `activateRequested` for the `nameOnSaveDialog` to piggyback on, but did not have much luck.

**Comment by jess-x** [comment]
> Question: as the list of files is store in an array activatedDocs, what will happen if the user refreshes (reloads) his page?

From what I understand, the any would be emptied when a refresh/reload happens.

**Comment by echarles** [comment]
@jess-x I have tested this branch and sometimes a persisted `Untitled.ipynb` still trigger a popup. I have also encountered a case where no pop-up was triggered but was not able to replicate.

Maybe this PR fixes a specific issues, but other issues are still showing up to me.

Even if we use the notebook metadata to bring more context (like a field `named: true`) that would help to have a fix, we would still miss all the other file formats like markdown...

At some point, if we can converge to a stable solution, we may need to revert this feature.

**Comment by krassowski** [comment]
> I have also encountered a case where no pop-up was triggered but was not able to replicate.

I saw that too and I was unable to replicate this reliably.

Implementation wise, I think that the anonymous function that gets connected to `widget.context.saveState` should get disconnected after rename or user confirming save without rename.

**Comment by jess-x** [comment]
> At some point, if we can not converge to a stable solution, we may need to revert this feature.

I agree, we could revert this feature since it looks like we might not be able to converge to a comprehensive fix before the RC release

cc @blink1073

**Comment by jtpio** [comment]
> I agree, we could revert this feature since it looks like we might not be able to converge to a comprehensive fix before the RC release

There are also a couple of attributes such as `renamed` that were added to the public API. And it looks like they might not be needed for the feature to work?

https://github.com/jess-x/jupyterlab/blob/88b8c145ea95fd3654987d3a490a9b996905dfb2/packages/services/src/contents/index.ts#L110-L115

I wanted to have a closer look at this to see if we could remove them before 3.1 so we don't break the API later if we decide to remove them after. But not sure there will be enough time for this.

**Comment by echarles** [comment]
> Put another way, we could indeed revert that feature for 3.1, and bring it back after the release for 3.2 or 4.0.

@jess-x Worth trying to open a PR to revert and see if it can get merged before 3.1

We can take over that feature just after, now that we have identified the trade-offs to deal with in the difficult cases.

**Comment by jess-x** [comment]
Name File dialog feature dumped to later release

**Pull request closed without merging**